### PR TITLE
Create Order - Add cart rule to cart - Apply cart rules first before calculate the cart totals

### LINKED ISSUE #25579: BO - Add new Order page - Summary block - Total taxes  & Total (Tax excl.) & Total (Tax incl.) are not updated when adding a voucher without code
#### Describe the bug

In the BO > Add new Order page > in the summary block, the Total taxes  & Total (Tax excl.) & Total (Tax incl.) are not updated when adding a voucher without code.
It is just an ajax issue.

#### Steps to Reproduce

1. Go to BO > Orders > Add new Order
2. Choose a customer and choose a product
3. Click on Add new voucher > Fill fields of a voucher without code for example discount amount = 1euro 
4. Scroll down > Check Total taxes  & Total (Tax excl.) & Total (Tax incl.) in the summary
4. See an error
image
6. if you try to change any field (change currency / add new product / change language / increase & decrease quantity) => ok
image

Also, if the cart rule has a gift => the gift is not added to the list of the products

In the PS1769 => ok
image

#### Additional information

* PrestaShop version: 1.7.8.0rc1build2 / 1776 / 1775
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.8.x
| Description?      | See below
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #25579 
| How to test?      | See #25579 
| Possible impacts? | -

## Description (copy of https://github.com/PrestaShop/PrestaShop/issues/25579#issuecomment-914269727)

Before, The modal was adding the cartRule to the cart in a fancyBox : we load the addCartRule page into an iframe. Nothing says that we want to link the cartRule added to the current cart
https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/admin-dev/themes/new-theme/js/pages/order/create/create-order-page.js#L185

The totals are calculated first and then we get the discounts. When we get the discounts, we call `getCartRules` with `autoAdd = true`. This query adds all applicable cartRules to the cart given. In this way, the totals are wrong and the discounts correct
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Cart.php#L4010
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Cart.php#L4047

One solution to avoid this is a BIG bandage : we first get the cartRules with autoAdd = true. That will apply all applicable cartRules to the cart and then we calculte the totals and the discounts

The ideal solution would be to have a specific modal to add cartRules to current cart. Doing this we will be able to perform
1 - create the cart Rule
2 - Try to apply created cart rule to the current cart (I say TRY because if the cart doesnt match the cartRule rules, it won't be applied)
This solution will also allow us to use the commands AddCartRuleCommand and AddCartRuleToCartCommand instead of the legacy controller like it is now 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25802)

### CODE REVIEW

**Review by atomiix — DISMISSED** [review]

File: classes/Cart.php
Comment (matks) [inline classes/Cart.php]: Maybe
```suggestion
        CartRule::autoAddToCart(...);
```
would make it more obvious that we do this, at this line, in order to apply to cart the cart rules?

**Review by matks — DISMISSED** [review]
Code is OK for me. I add a suggestion.

**Review by Progi1984 — COMMENTED** [review]
@sowbiba Hi Could you add behat test for checking this behavior change, please ?

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
I agree with @matks , it's not obivous and can be confusing, people from the future will maybe refactor it, remove the "useless" variable thinking it's not needed.
Using autoAddToCart will be better than adding a long comment :/

**Comment by sowbiba** [comment]
> I agree with @matks , it's not obivous and can be confusing, people from the future will maybe refactor it, remove the "useless" variable thinking it's not needed.
> Using autoAddToCart will be better than adding a long comment :/

Ok, but doing this will make me change the getCartRules using autoAdd to false to avoid useless queries. And the comment wil be still relevant to explain why we do this at this time

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
I'm not sure your test is relevant. I tried with and without your change, and test is still ok :sweat_smile:

**Comment by sowbiba** [comment]
Oh, forgot to check the most important part : the totals

File: tests/Integration/Behaviour/Features/Context/CartFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/CartFeatureContext.php]:
```suggestion
    public function assertCartRuleIsAppliedToCart(string $cartRuleReference): void
```

File: tests/Integration/Behaviour/Features/Context/CartFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/CartFeatureContext.php]:
```suggestion
    public function checkCartToOrderValues(TableNode $table): void
```

File: tests/Integration/Behaviour/Features/Context/CartFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/CartFeatureContext.php]: Please define the return type

File: tests/Integration/Behaviour/Features/Scenario/Cart/CartToOrder/cart_to_order.feature
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Scenario/Cart/CartToOrder/cart_to_order.feature]:
```suggestion
    And email sending is disabled
    And there is a product in the catalog named "product1" with a price of 19.812 and 1000 items in stock
```

File: tests/Integration/Behaviour/Features/Scenario/Cart/CartToOrder/cart_to_order.feature
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Scenario/Cart/CartToOrder/cart_to_order.feature]:
```suggestion
    And there is a cart rule named "cartrule1" that applies an amount discount of 5.0 with priority 1, quantity of 1000 and quantity per user 1000
    And there is a customer named "customer1" whose email is "fake@prestashop.com"
```

**Review by Progi1984 — COMMENTED** [review]
@sowbiba Thank you for your contribution. I have brought you some comments.

File: classes/Cart.php
Comment (matks) [inline classes/Cart.php]: Did you notice that a lot of functions in Cart.php do
```
$context = Context::getContext()->cloneContext();
$context->cart = $this;
```
Maybe this is in order to avoid side effects when doing
```
$context->cart = $this;
```

File: tests/Integration/Behaviour/Features/Context/CartRuleFeatureContext.php
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/CartRuleFeatureContext.php]: After this change, any Behat tests that uses `createCartRule()` will create Cart Rules that have no code and, as you mentioned, cart rules with no codes are automatically applied when eligible.

I'm a bit afraid of the consequences of this change. 🤷‍♂️ However CI seems green ... so maybe there's no issue.

**Review by matks — COMMENTED** [review]
Code seems OK to me. I just have one thing that worries me.

**Review by atomiix — DISMISSED** [review]

File: classes/Cart.php
Comment (PierreRambaud) [reply classes/Cart.php]: If you do this you're not cloning anymore, the reference is the same.

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
I don't think it should be done here.
The aim of this method is to retrieve the raw summary details, not apply cart rules.

File: classes/Cart.php
Comment (jolelievre) [reply classes/Cart.php]: The cart reference is the same, but not the context one That's the idea, so you can perform "changes" on this new context with fewer issues I guess... The context was already a problem back in the days ^^

File: classes/Cart.php
Comment (jolelievre) [inline classes/Cart.php]: I'm not sure it's the best way 🤔  But I understand your issue that the cart rules have not been updated. Maybe you have a simpler solution here if you call:

```php
$cartRules = $this->getCartRules();
```

right here instead of L4051, then they also should be automatically added.

Then on L4051 you can do:

```php
'discounts' => array_values($cartRules),
```

Comment (jolelievre) [reply classes/Cart.php]: But as you mentioned in the issue, this is just a big ugly patch, the problem is not that the `getRawSummaryDetails` doesn't add the cart rules when expected (it'snot its job). The problem comes from the fact that the created cart rule in the create order page was never added to the cart in the first place.

You should fix the problem at the source (adding a cart rule correctly to a cart), not at the output.

Comment (jolelievre) [reply classes/Cart.php]: There are two ways to fix this correctly:
- the URL in the iframe should not be a generic independent "Cart rule form", but a dedicated controller "Add cart rule to a cart", that handles to add the cart rule to a cart right after it's been created
- keep using the generic cart rule form, but then you ned to get back the generated cart rule ID, and add it to the cart via `this.cartRuleManager.addCartRuleToCart(cartRuleId, <number> this.cartId);`

File: tests/Integration/Behaviour/Features/Context/CartFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/CartFeatureContext.php]: Nice check!!

File: tests/Integration/Behaviour/Features/Context/CartRuleFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/CartRuleFeatureContext.php]: I don't understand the use of this extra `$code` parameter since you don't use in your endpoint any way

Ok it's empty string instead of null, but does it really changes the behavior? That would mean all present cart rules have been created with a code and we didn't realize it?

**Review by jolelievre — CHANGES_REQUESTED** [review]
I believe you are fixing the symptom instead of the root cause

File: admin-dev/themes/new-theme/js/pages/order/create.js
Comment (matks) [inline admin-dev/themes/new-theme/js/pages/order/create.js]: I think this sentence can be improved 😅  as I dont fully understand it.

```suggestion
    console.log('Error: Cannot apply created cart rule to the current cart as orderPageManager is not available');
```

File: admin-dev/themes/default/template/controllers/cart_rules/form.tpl
Comment (matks) [inline admin-dev/themes/default/template/controllers/cart_rules/form.tpl]: Maybe indent like below?
```suggestion
		    window.parent.order_create.addCreatedCartRuleToCart({$created_cart_rule});
```

**Review by matks — DISMISSED** [review]
I approve this change as _acceptable_ for a patch 😄 

it was not easy to find something that does not reinvent the wheel and handle current issue 👍

File: admin-dev/themes/default/template/controllers/cart_rules/form.tpl
Comment (jolelievre) [inline admin-dev/themes/default/template/controllers/cart_rules/form.tpl]: Can you add a comment on this whole block (before the `if isset($refresh_cart)`) to explain the purpose of this code It is related to the "Create order page" which can create cart rule via a fancy box and needs to update its content via JS code calls

Comment (jolelievre) [reply admin-dev/themes/default/template/controllers/cart_rules/form.tpl]: This will be important when we migrate this page, to understand why this piece of code is present here and needs to be migrated or, even better, modified with a cleaner solution (dedicated form, or smarter JS code that is able to fetch the create CartRuleId)

File: controllers/admin/AdminCartRulesController.php
Comment (jolelievre) [inline controllers/admin/AdminCartRulesController.php]:
```suggestion
                $this->context->smarty->assign('created_cart_rule_id', $cart_rule->id);
```

Comment (jolelievre) [reply controllers/admin/AdminCartRulesController.php]: I think both variables have the same effect actually So they should be merged into one You can replace `refresh_cart` with only `created_cart_rule_id` and in the template you can only check for `created_cart_rule_id` presence and call the `addCreatedCartRuleToCart` js function only

The internally `addCreatedCartRuleToCart` can call the `refreshCart` function this way you are sure, the add cart rule action has been performed fully before the cart is refreshed and it reduces the quantity of code necessary in the smarty template

File: admin-dev/themes/default/template/controllers/cart_rules/form.tpl
Comment (jolelievre) [reply admin-dev/themes/default/template/controllers/cart_rules/form.tpl]: As suggested in the cart controller, these two functions call and two variables could be merged into one because it's the same use case/behaviour

File: admin-dev/themes/new-theme/js/pages/order/create.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/order/create.js:72]:
```suggestion
  orderPageManager.addCreatedCartRuleToCart(cartRuleId);
  orderPageManager.refreshCart();
```

**Review by jolelievre — COMMENTED** [review]
I agree with this solution as a temporary one Ideally the clean solution would be to have a dedicated controller that allows to create a cart rule and add it to the cart automatically (like an overlay controller of the cart rule creation action).

But for now this is enough, although I made. few suggestions to simplify and clarify the code, and a few comments would be nice so that we understand the purpose of this code when the form is migrated in the future (to avoid forgetting this feature)

And finally, you added some new endpoints in the behat tests, but they are never used I believe this is a leftover from your initial modification? If this new end point is not used anymore it should be cleaned from this PR.

File: admin-dev/themes/default/template/controllers/cart_rules/form.tpl
Comment (Progi1984) [inline admin-dev/themes/default/template/controllers/cart_rules/form.tpl]: It seems the indentation is broken.

File: admin-dev/themes/new-theme/js/pages/order/create.js
Comment (Progi1984) [reply admin-dev/themes/new-theme/js/pages/order/create.js:72]: Why not applying this suggestion @sowbiba ?

**Review by Progi1984 — COMMENTED** [review]
@sowbiba Thank you for your contribution. Some comments.

File: admin-dev/themes/new-theme/js/pages/order/create.js
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/order/create.js:72]: addCreatedCartRuleToCart already perform a refreshCart

File: admin-dev/themes/default/template/controllers/cart_rules/form.tpl
Comment (tegbessou) [inline admin-dev/themes/default/template/controllers/cart_rules/form.tpl]: Indent problem on this line and following

File: admin-dev/themes/new-theme/js/pages/order/create.js
Comment (tegbessou) [inline admin-dev/themes/new-theme/js/pages/order/create.js:69]: It's normal to let console.log?

**Review by tegbessou — DISMISSED** [review]
Two small returns, but no need to block I think

File: admin-dev/themes/new-theme/js/pages/order/create.js
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/order/create.js:69]: It's the same in other methods of this file

I guess it's for helping devs to debug without throwing an error

Comment (atomiix) [reply admin-dev/themes/new-theme/js/pages/order/create.js:69]: @tegbessou if I'm not mistaken, `console.log`s are removed at build time in production environment.

**Comment by atomiix** [comment]
I think tests are missing?

File: admin-dev/themes/default/template/controllers/cart_rules/form.tpl
Comment (Progi1984) [inline admin-dev/themes/default/template/controllers/cart_rules/form.tpl]: Set this comment between Smarty comments tags

**Review by Progi1984 — COMMENTED** [review]
@sowbiba A last comment and it's ok for me.

**Comment by sowbiba** [comment]
> I think tests are missing?

🤔 The fix is in JS. Do you see any relevant test I can add to cover this feature ?

**Review by atomiix — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]
Thank you @sowbiba

**Comment by PierreRambaud** [comment]
> > I think tests are missing?
> 
> thinking The fix is in JS. Do you see any relevant test I can add to cover this feature ?

If the fix is in JS, why change php codes? :thinking: 
https://github.com/PrestaShop/PrestaShop/pull/25802/files#diff-71f172be7472328a8a53debe33853f5fa6611123ef8f411c393123e013f9fda8R117

**Review by PierreRambaud — COMMENTED** [review]
I think some tests are missing, or PHP parts shouldn't be changed without any reasons :thinking:

**Comment by enginteknoloji** [comment]
1.7.8.3 After the installation, a different Theme is Installed and the Installed Theme is run again.. When the product is added to the cart from the Installed Theme. There is no problem in adding the product in the basket. However, the product in the basket remains in the basket after clicking the Remove button. After the page is refreshed, the product in the basket is deleted.. Can you make this correction in all the themes?(https://hepsioffis.com/women/2-9-brown-bear-printed-sweater.html#/1-boyut-s)

**Review by jolelievre — APPROVED** [review]

**Comment by jolelievre** [comment]
Hi @enginteknoloji 

I'm not sure this is the relevant place to discuss your bug, you are talking about themes which are only available in the Front Office, this PR aims at fixing a bug in the Back Office when creating an order from the BO. So I'm not sure it's related.

That said if you do have found a bug don't hesitate to [open a dedicated issue](https://github.com/PrestaShop/PrestaShop/issues/new/choose)

File: src/Adapter/Cart/CommandHandler/AddCartRuleToCartHandler.php
Comment (jolelievre) [inline src/Adapter/Cart/CommandHandler/AddCartRuleToCartHandler.php]: Hi @sowbiba 

we were wondering if this modification is still relevant? Initially, this PR fixed the issue via the command handler, then during the discussion, we went for a different approach that is handled via the BO front and JS.

But what about this modification? Is it still needed or is it a relic of the previous implementation? Or said differently should this be cleaned? or should we add some behat test to validate this modification is important and not modified later because it solves a specific issue?

**Review by jolelievre — CHANGES_REQUESTED** [review]
We have one last question for this one, so I prefer to block the PR until everything is clear

**Review by jolelievre — APPROVED** [review]
Thanks @sowbiba

**Review by Progi1984 — APPROVED** [review]

**Review by HanaRebaiQA — CHANGES_REQUESTED** [review]
Hello @sowbiba 

I cheked this PR. The issue is not fixed.
Here the steps : 
-Go to CONFIGURE >  Shop Parameters > Customer Settings 
-Go to Groups > Customer > Edit > Set the Price display method on `Tax excluded`
-Then, Go to BO > SELL > Orders > Orders > Add new order
-Click on `Add new Voucher ` > Fill fields of a voucher 
-In ACTIONS > set an Amount = 1 euro for example and set `Tax Included` > Save
-Add a new Voucher with Tax Included 
Or
-In ACTIONS > set an Amount = 1 euro for example and set `Tax Excluded` > Save
-Add a new Voucher with Tax Included 
-See error > The Total (Tax excl.) is incorrect ==> does not take into account the total of the vouchers
image

After order confirmation
image

See the attached videos 

https://user-images.githubusercontent.com/92912932/156786926-c9a02ee8-c469-4e29-afdb-ad51edd71b03.mp4


https://user-images.githubusercontent.com/92912932/156790388-d32ae02e-a79b-4039-b8a7-a50f3731bfe5.mp4

https://user-images.githubusercontent.com/92912932/156790572-000fc7bc-96c3-49f5-ab5d-673909ee6f1e.mp4


https://user-images.githubusercontent.com/92912932/156791143-60587aaf-d33a-477a-a8be-3fe7450aa952.mp4

Could yous please check and feedback ?

Thanks!

**Comment by sowbiba** [comment]
Maybe I missed something but, the tests are OK for me

https://user-images.githubusercontent.com/7426640/157649911-35a5d21a-bd22-47af-a13b-53b54c1b3aaf.mp4

**Comment by sowbiba** [comment]
Saw with @HanaRebaiQA 
For the first order, no problem. For the 2nd order created through this page, the first cartRule is computed well but the second one is ignored. Weird.
I'm investigating to understand what happens

**Comment by matks** [comment]
@sowbiba did you find some more answers 😉 ?

**Comment by matks** [comment]
Hi @sowbiba it seems very hard for you to continue this issue. With no activity for more than 30 days, I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale. You can reopen the PR if you think you can go forward 😉 

Thanks!

**Pull request closed without merging**
